src/kinsol/kinsol.c

```diff
@@ -3010,50 +2975,148 @@ static int KINFP(KINMem kin_mem)
+        temp                 = c * a + s * b;
+        R[j * depth + i + 1] = -s * a + c * b;
+        R[j * depth + i]     = temp;
+      }
+    }
+    N_VLinearSum(c, Q[i], s, Q[i + 1], kin_mem->kin_vtemp2);
+    N_VLinearSum(-s, Q[i], c, Q[i + 1], Q[i + 1]);
+    N_VScale(ONE, kin_mem->kin_vtemp2, Q[i]);
+  }
+
+  /* Shift R to the left by one. */
+  for (int i = 1; i < depth; i++)
+  {
+    for (int j = 0; j < depth - 1; j++)
+    {
+      R[(i - 1) * depth + j] = R[i * depth + j];
+    }
+  }
+
+  /* If ICWY orthogonalization, then update T */
+  if (kin_mem->kin_orth_aa == KIN_ORTH_ICWY)
+  {
+    if (dotprodSB)
+    {
+      if (depth > 1)
+      {
+        for (int i = 2; i < depth; i++)
+        {
+          N_VDotProdMultiLocal(i, Q[i - 1], Q,
+                               kin_mem->kin_T_aa + (i - 1) * depth);
+        }
+        N_VDotProdMultiAllReduce(depth * depth, Q[depth - 1], kin_mem->kin_T_aa);
+      }
+      for (int i = 1; i < depth; i++)
+      {
+        kin_mem->kin_T_aa[(i - 1) * depth + (i - 1)] = ONE;
+      }
+    }
+    else
+    {
+      kin_mem->kin_T_aa[0] = ONE;
+      for (int i = 2; i < depth; i++)
+      {
+        N_VDotProdMulti(i - 1, Q[i - 1], Q, kin_mem->kin_T_aa + (i - 1) * depth);
+        kin_mem->kin_T_aa[(i - 1) * depth + (i - 1)] = ONE;
+      }
+    }
+  }
+
+  return KIN_SUCCESS;
+}
+
 static int AndersonAcc(KINMem kin_mem, N_Vector gval, N_Vector fv, N_Vector x,
                        N_Vector xold, long int iter, sunrealtype* R,
                        sunrealtype* gamma)
 {
   int retval;
-  long int i_pt, i, j, lAA;
-  long int* ipt_map;
+  long int lAA;
   sunrealtype alfa;
   sunrealtype onembeta;
-  sunrealtype a, b, temp, c, s;
-  sunbooleantype dotprodSB = SUNFALSE;
 
   /* local shortcuts for fused vector operation */
   int nvec        = 0;
   sunrealtype* cv = kin_mem->kin_cv;
   N_Vector* Xv    = kin_mem->kin_Xv;
 
-  /* local dot product flag for single buffer reductions */
-  if ((kin_mem->kin_vtemp2->ops->nvdotprodlocal ||
-       kin_mem->kin_vtemp2->ops->nvdotprodmultilocal) &&
-      kin_mem->kin_vtemp2->ops->nvdotprodmultiallreduce)
-  {
-    dotprodSB = SUNTRUE;
-  }
-
-  ipt_map = kin_mem->kin_ipt_map;
-  i_pt    = iter - 1 - ((iter - 1) / kin_mem->kin_m_aa) * kin_mem->kin_m_aa;
-
   /* Compute residual F(x) = G(x_old) - x_old */
   N_VLinearSum(ONE, gval, -ONE, xold, fv);
 
   if (iter > 0)
   {
+    /* If we've filled the acceleration subspace, start recycling */
+    if (kin_mem->kin_current_depth == kin_mem->kin_m_aa)
+    {
+      /* Move the left-most column vector (oldest value) to the end so it gets
+         overwritten with the newest value below. */
+      N_Vector tmp_dg = kin_mem->kin_dg_aa[0];
+      N_Vector tmp_df = kin_mem->kin_df_aa[0];
+      for (long int i = 1; i < kin_mem->kin_m_aa; i++)
+      {
+        kin_mem->kin_dg_aa[i - 1] = kin_mem->kin_dg_aa[i];
+        kin_mem->kin_df_aa[i - 1] = kin_mem->kin_df_aa[i];
+      }
```

Comment:
Optional suggestion: use `memmove`

Reply:
I'll add this to the issue for future enhancements